File: bin/adding_method/messi_add_method/method-template/modules/method/predict/resources/usr/bin/method_predict.py
#!/usr/bin/env python

"""
This is a {{ method|lower }} prediction script, loads torch like models and make predictions
Output a summary table with test probability in it.

Usage:
  {{ method|lower }}_predict.py [options]

Options:
  -h --help                   Show this message
  --model_path=MODEL          Path to trained model on specific fold [default: empty]
  --test_path=TEST_PATH       Path that contains the input for testing [default: empty]
  --metadata_path=META        Path to the additional metadata like sample name, response [default: empty]
  --label=LABEL               Label of dataset and fold iteration [default: empty]
  --method_name=METHOD_NAME   Method name ran [default: {{ method|lower }} ]
"""

# TODO: The docopt help message above can be better described or reformatted

from docopt import docopt
import pickle
import pandas as pd

def predict_test(model, test_data):
    # TODO: You could have this function in this script or place it elsewhere
    # Takes the model and predict on the testdata
    # testdata should be in N x p_i format, where N is total number of observations, 
    # p_i is number of features in i views of omics
    # Should return a n x 1 dimensional vector of probabilities in positive class of binary classification
    raise NotImplementedError
    

def generate_result_table(predicted, metadata_path, method_name, decimals=3):
    # First read in the previous metadata table
    df = pd.read_csv(metadata_path)
    # Append the method name and predicted probabilities
    df["phat"] = predicted
    df["method_name"] = method_name
    # reorder it to desired format
    desired_order = ["sample_name", "y", "phat", "method_name", "dataset"]
    df = df[desired_order]
    # Sample name as to be always string, sometimes it could come in as integer
    df["sample_name"] = df["sample_name"].astype('string')
    # Round phat
    df['phat'] = df['phat'].apply(lambda x: round(x, decimals))
    return df

# TODO: Implement this main function to be taking a fold-specific model fit on train data
#       and evaluate/predict on the fold-specific test data
def main(model_path, test_path, metadata_path, label, method_name):
  # Load model
  with open(model_path, 'rb') as mf:
      model = pickle.load(mf)
  # Load test input
  with open(test_path, 'rb') as tf:
      test_data = pickle.load(tf)
  # TODO: Need to implement predict_test and generate_result_table accordingly
  predicted = predict_test(model=model, test_data=test_data)
  result_table = generate_result_table(predicted=predicted, metadata_path=metadata_path, method_name=method_name)
  # Write the result table to csv
  result_file = f"{label}-result.csv"
  result_table.to_csv(result_file, index=False, header=True)
  return result_table

# Execute it here
if __name__ == '__main__':
  # Parse docopt
  args = docopt(__doc__)
  # Execute the main function
  main(
    model_path=args['--mode_path'], 
    test_path=args['--test_path'], 
    metadata_path=args['--metadata_path'],
    label=args['--label'],
    method_name=args['--method_name']
  )

File: bin/adding_method/messi_add_method/method-template/modules/method/predict/resources/usr/bin/method_predict.R
#!/usr/bin/env Rscript

# Script to predict {{ method|lower }} 
doc <- "This script is to make predictions on test data of particular fold, 
using a model trained from the {{ method|lower }} method.

Output type is a path containing the predicted probabilities.

Usage:
  {{ method|lower }}_predict.R [options]

Options:
  --model_path=MODEL_PATH     Path to read the model [default: null]
  --test_path=TEST_PATH       Path containing test data [default: null]
  --label=LABEL               Label of id and fold of data [default: data-fold_i]
  --method_name=METHOD_NAME   Method name input from upstream [default: {{ method|lower }}]
  --output_ext=EXT            Extension of output table to save [default: csv]
"
# Load libraries
library(tibble)
library(magrittr)
library(dplyr)
library(here)
# Source scripts
# This is a helper script to help combine and store results for downstream comparison
source(here("bin/wrangling/get_result_table.R"))
# Parse docopt
opt <- docopt::docopt(doc)
# This is the function logic
# TODO: add suitable args for your method
main <- function(model_path, test_path, label, output_ext, method) {
  # Load model (fit wit train data) and test data of particular fold split
  model <- readRDS(model_path)
  test_data <- readRDS(test_path)
  # Verbose input for debug
  message("\nRead model from ", model_path, "\n")
  message("\nRead test data from ", test_path, "\n")
  # TODO: Implement your logic of getting predicted probabilities of positive class in
  # a binary classification problem
  # Predict and get result
  pred_probs <- predict(model, test_data) # THIS WILL YOU ERROR, so implement this

  # Make sure you could have rownames for identifying different observations/patients
  probs_df <- pred_probs %>%
              as.data.frame() %>%
              rownames_to_column(var="sample_name") %>%
              as_tibble()
  # TODO: might need to transform your output a little to suit this helper function
  # that wrangles and makes it ready for downsteram comparison
  # Merge to summary table
  result_table <- get_result_table(probs=probs_df, label=label, 
                                  method_name=method,
                                  test_data=test_data
                                  )
  
  # Write to files
  message("\nSaving as ", output_ext, " format\n")
  result_file <- paste(label, paste0("result_table", ".", output_ext), sep="-")
  # Save to disk
  write.csv(result_table, result_file, row.names = FALSE)
  return(pred_obj)
}


# Call the function here
main(model_path=opt$model_path, 
     test_path=opt$test_path, 
     label=opt$label,
     output_ext=opt$output_ext,
     method=opt$method
)

message("Done")